File: example/python/ising.py
# coding:utf-8

import os, sys
currentdir = os.path.dirname(os.path.realpath(__file__))
sys.path.append(os.path.join(currentdir, '../../src'))

import math
import numpy as np
from alps import alea

def xy2i(L, x, y):
    return L * y + x

def i2x(L, i):
    return i % L

def i2y(L, i):
    return i // L

def left(L, i):
  x = (i2x(L, i) + L - 1) % L
  y = i2y(L, i)
  return xy2i(L, x, y)

def right(L, i):
  x = (i2x(L, i) + 1) % L
  y = i2y(L, i)
  return xy2i(L, x, y)

def up(L, i):
  x = i2x(L, i)
  y = (i2y(L, i) + 1) % L
  return xy2i(L, x, y)

def down(L, i):
  x = i2x(L, i)
  y = (i2y(L, i) + L - 1) % L
  return xy2i(L, x, y)

L = 8
N = L * L
T = 2.269185314213022
nsamples = (1 << 12)
ntherm = nsamples // 8

spins = np.ones(N, dtype=int)

np.random.seed(1234)

ene = alea.autocorr()
ene2 = alea.batch_2()
mag2 = alea.autocorr()

for m in range(ntherm + nsamples):
    for i in range(N):
        dE = 2.0 * spins[i] * (spins[left(L, i)] + spins[right(L, i)] + spins[up(L, i)] + spins[down(L, i)])
        if (np.random.random() < math.exp(-dE / T)):
            spins[i] = -1 * spins[i]
    if (m >= ntherm):
        E = 0.0
        m = 0.0;
        for i in range(N):
            E -= 1.0 * spins[i] * (spins[right(L, i)] + spins[up(L, i)])
            m += 1.0 * spins[i]
        E /= N
        m /= N
        ene.add(E)
        ene2.add(E, E * E)
        mag2.add(m * m)

print("L =", L)
print("T =", T)
print("Ns =", nsamples)
  
ene = ene.finalize()
print("<E> = {} +/- {}".format(ene.mean(), ene.stderror()))
print("tau(E) =", ene.tau())

ene2 = ene2.finalize()
ene2 = ene2.variance()
print("<C> = {} +/- {}".format(N * ene2.mean(0) / (T * T), N * ene2.stderror(0) / (T * T)))

mag2 = mag2.finalize()
print("<m^2> = {} +/- {}".format(mag2.mean(), mag2.stderror()))
print("tau(m^2) =", mag2.tau())

File: example/cxx/ising.cpp
#include <cmath>
#include <iostream>
#include <random>
#include <vector>
#include <alps/alea.hpp>

inline int xy2i(int L, int x, int y) { return L * y + x; }

inline int i2x(int L, int i) { return i % L; }

inline int i2y(int L, int i) { return i / L; }

inline int left(int L, int i) {
  int x = (i2x(L, i) + L - 1) % L;
  int y = i2y(L, i);
  return xy2i(L, x, y);
}

inline int right(int L, int i) {
  int x = (i2x(L, i) + 1) % L;
  int y = i2y(L, i);
  return xy2i(L, x, y);
}

inline int up(int L, int i) {
  int x = i2x(L, i);
  int y = (i2y(L, i) + 1) % L;
  return xy2i(L, x, y);
}

inline int down(int L, int i) {
  int x = i2x(L, i);
  int y = (i2y(L, i) + L - 1) % L;
  return xy2i(L, x, y);
}

int main() {
  int L = 8;
  int N = L * L;
  double T = 2.269185314213022;
  int nsamples = (1 << 12);
  int ntherm = nsamples / 8;

  std::vector<int> spins(N, 1);

  std::mt19937 engine(0);
  std::uniform_real_distribution<> uniform;

  alps::alea::autocorr_acc<double> ene;
  alps::alea::batch_acc<double> ene2(2);
  alps::alea::autocorr_acc<double> mag2;

  for (int m = 0; m < (ntherm + nsamples); ++m) {
    for (int i = 0; i < N; ++i) {
      double dE = 2.0 * spins[i] * (spins[left(L, i)] + spins[right(L, i)] + spins[up(L, i)] + spins[down(L, i)]);
      if (uniform(engine) < std::exp(-dE / T))
        spins[i] = -spins[i];
    }
    if (m >= ntherm) {
      double E = 0.0;
      double m = 0.0;
      for (int i = 0; i < N; ++i) {
        E -= spins[i] * (spins[right(L, i)] + spins[up(L, i)]);
        m += spins[i];
      }
      E /= N;
      m /= N;
      ene << E;
      ene2 << alps::alea::column<double>{ E, E * E };
      mag2 << m * m;
    }
  }

  std::cout << "L = " << L << std::endl;
  std::cout << "T = " << T << std::endl;
  std::cout << "Ns = " << nsamples << std::endl;
  
  auto ene_corr = ene.finalize();
  std::cout << "<E> = " << ene_corr << std::endl;
  std::cout << "tau(E) = " << ene_corr.tau() << std::endl;

  auto ene2_batch = ene2.finalize();
  std::function<double(double, double)> f = [](double x0, double x1) { return x1 - x0 * x0; };
  auto ene2_prop = alps::alea::transform(alps::alea::jackknife_prop(), alps::alea::make_transformer(f), ene2_batch);
  std::cout << "C = " << N * ene2_prop.mean() / (T * T) << " +/- " << N * ene2_prop.stderror() / (T * T) << std::endl;

  auto mag2_corr = mag2.finalize();
  std::cout << "<m^2> = " << mag2_corr << std::endl;
  std::cout << "tau(m^2) = " << mag2_corr.tau() << std::endl;
}
